[HLSL][RootSignature] Implement parsing of a DescriptorTable with empty clauses

inbelic opened the pull request:
- defines the Parser class and an initial set of helper methods to support consuming tokens. functionality is demonstrated through a simple empty descriptor table test case
- defines an initial in-memory representation of a DescriptorTable
- implements a test harness that will be used to validate the correct diagnostics are generated. it will construct a dummy pre-processor with diagnostics consumer to do so

Implements the first part of https://github.com/llvm/llvm-project/issues/126569

inbelic committed:
[NFC][HLSL][RootSignature] add spelling of tokens to tokenkind.def

Commit:
[HLSL][RootSignature] Implement parsing of an empty DescriptorTable

- defines the Parser class and an initial set of helper methods to
support consuming tokens. functionality is demonstrated through a simple
empty descriptor table test case
- defines an initial in-memory representation of a DescriptorTable
- implements a test harness that will be used to validate the correct
diagnostics are generated. it will construct a dummy pre-processor with
diagnostics consumer to do so

Commit:
implement parsing of empty DescriptorTableClause

Commit:
self-review: docs

Commit:
review: clang-format

Commit:
self-review: fix up comment

Commit:
review: fix build issue on api change

inbelic commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -0,0 +1,110 @@
+  //  if (Parse.*())
+  //    return true;
+  //
+  //   - All methods return true if a parsing error is encountered. It is the
```
Is it clear here that methods is the `Parse.*` methods?

bogner replied:
Yes, this seems clear.

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,188 @@
+  return TokenString;
+}
+
+// Parser Definitions
```
nit: can remove

joaosaffran commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,188 @@
+
+bool RootSignatureParser::Parse() {
+  // Iterate as many RootElements as possible
+  while (TryConsumeExpectedToken(TokenKind::kw_DescriptorTable)) {
```
nit: Something that makes me confuse with that approach is, how will that look once you add more RootElements? Will that become something like:
```C++
while (
TryConsumeExpectedToken(TokenKind::kw_DescriptorTable) || 
TryConsumeExpectedToken(TokenKind::kw_RootDescriptor) || 
TryConsumeExpectedToken(TokenKind::kw_RootConstant)) {
...
}
```

My expectation for that code would be something like: 
```C++
do{
// Parsing logic here
} while(TryConsumeExpectedToken(TokenKind::pu_comma));
```

inbelic replied:
Correct, it will look like how you mentioned in the other comment:
```
while (TryConsumeExpectedToken({TokenKind::kw_DescriptorTable, TokenKind::kw_RootConstant, ... }) {
...
}
```

joaosaffran commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,188 @@
+    return true;
+
+  // Iterate as many Clauses as possible
+  while (TryConsumeExpectedToken({TokenKind::kw_CBV, TokenKind::kw_SRV,
```
nit: I imagine the above loop would be something like this, once you add more Root Elements?

joaosaffran commented on llvm/include/llvm/Frontend/HLSL/HLSLRootSignature.h:
```diff
@@ -0,0 +1,44 @@
+};
+
+// Models RootElement : DescriptorTable | DescriptorTableClause
+using RootElement = std::variant<DescriptorTable, DescriptorTableClause>;
```
I feel like the `DescriptorTableClause` should be a member of `DescriptorTable` and not a RootElement, since those are connected in the metadata representation, it should be easier that way, IMHO.

inbelic replied:
I agree that would be intuitive, but I would argue that each of these clauses will have their own metadata node that this representation maps well to.

There is also an implementation hurdle to have these as a vector on `DescriptorTable` which is that we need to explicitly allocate all the memory onto the `AST`. Practically this means that after this function returns in `SemaHLSL` we would need to go through and reallocate all these vectors as `ArrayRef`.

Or maybe we can pass down an allocator? I am not sure, either way I think it would introduce complexity rather than remove it

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
Anddd forgot to add the liscence again... Will do when addressing other comments

bogner commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -0,0 +1,110 @@
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the ParseHLSLRootSignature interface.
```
The interface is called "RootSignatureParser", not "ParseHLSLRootSignature"

bogner commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -0,0 +1,110 @@
+  /// error is encountered or the end of the buffer.
+  ///
+  /// Returns true if a parsing error is encountered.
+  bool Parse();
```
Since this is all new code we should use the LLVM naming conventions for methods of starting with lower case, rather than matching the legacy conventions elsewhere in the `Parse` library. https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

bogner commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,188 @@
+namespace clang {
+namespace hlsl {
+
+static std::string FormatTokenKinds(ArrayRef<TokenKind> Kinds) {
```
You can simplify this while also making it more efficient by implementing a `TokenFormatter` class with an `operator<<`  for `DiagnosticBuilder`. The class should be okay to hold an `ArrayRef` as a member since it will always be used as a temporary, like `DB << TokenFormatter(Tokens);`

inbelic replied:
Updated to define the `const operator<<` for `TokenKind` and removed the api to use `consumeExpectedToken(ArrayRef)`

bogner commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,188 @@
+      llvm_unreachable("Switch for consumed token was not provided");
+    }
+
+    if (Error)
+      return true;
+
```
Why set `Error` rather than simply returning true from within the case statements?

bogner commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,188 @@
+  DescriptorTableClause Clause;
+  switch (CurToken.Kind) {
+  default:
+    break; // Unreachable given Try + assert pattern
```
Use `llvm_unreachable` when something is unreachable. Arguably you could do that instead of having the explicit assert at the top of the function, as it should have the same effect in this case.

bogner commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,188 @@
+
+bool RootSignatureParser::PeekExpectedToken(ArrayRef<TokenKind> AnyExpected) {
+  RootSignatureToken Result = Lexer.PeekNextToken();
+  return IsExpectedToken(Result.Kind, AnyExpected);
```
You could just use `std::find` (or it's range-aware llvm version) here if you like.

bogner reviewed (commented):
One comment saying you should rename all of the methods in this patch and a few small suggestions, but this is looking pretty close

Commit:
review: return true from case

Commit:
review: use unreachable

Commit:
review: conform to llvm method naming conventions

Commit:
review: use llvm::is_contained instead of custom function

Commit:
review: typos

Commit:
review: remove api of `consumeExpectedToken` for multiple token kinds

- using `consumeExpectedToken` with multiple expected tokens leads to
poor handling of diagnostic messaging and would largely be used in cases
where more specific errors could be done
- instead we should limit the api to just using `consumeExpectedToken`
with only a specific token and to only be used when it is very clear
that only one specific token should follow after another

Commit:
self-review: improve error messaging for expected end of parameters
